src/Xamarin.Android.Build.Tasks/Utilities/CecilExtensions.cs

```diff
@@ -0,0 +1,186 @@
+			var customAttrType = custom_attribute.Constructor.DeclaringType.FullName;
+			foreach (var t in MethodRegistrationAttributes) {
+				if (customAttrType == t)
+					return true;
+			}
+		}
+		return false;
+	}
+
+	public static IEnumerable<TAttribute> GetAttributes<TAttribute> (Mono.Cecil.ICustomAttributeProvider p, Func<CustomAttribute, TAttribute?> selector)
+			where TAttribute : class
+	{
+		return GetAttributes (p, typeof (TAttribute).FullName!, selector);
+	}
+
+	public static IEnumerable<TAttribute> GetAttributes<TAttribute> (Mono.Cecil.ICustomAttributeProvider p, string attributeName, Func<CustomAttribute, TAttribute?> selector)
+			where TAttribute : class
+	{
+		return p.GetCustomAttributes (attributeName)
+			.Select (selector)
+			.Where (v => v != null)
+			.Select (v => v!);
+	}
+
+	internal static IEnumerable<RegisterAttribute> GetTypeRegistrationAttributes (Mono.Cecil.ICustomAttributeProvider p)
+	{
+		foreach (var a in CecilExtensions.GetAttributes<RegisterAttribute> (p, a => CecilExtensions.ToRegisterAttribute (a))) {
+			yield return a;
+		}
+		foreach (var c in p.GetCustomAttributes ("Java.Interop.JniTypeSignatureAttribute")) {
+			var r = RegisterFromJniTypeSignatureAttribute (c);
+			if (r == null) {
+				continue;
+			}
+			yield return r;
+		}
+	}
+
+	internal static RegisterAttribute? RegisterFromJniTypeSignatureAttribute (CustomAttribute attr)
+	{
+		// attr.Resolve ();
+		RegisterAttribute? r = null;
+		if (attr.ConstructorArguments.Count == 1)
+			r = new RegisterAttribute ((string) attr.ConstructorArguments [0].Value, attr);
+		if (r != null) {
+			var v = attr.Properties.FirstOrDefault (p => p.Name == "GenerateJavaPeer");
+			if (v.Name == null) {
+				r.DoNotGenerateAcw = false;
+			} else if (v.Name == "GenerateJavaPeer") {
+				r.DoNotGenerateAcw = !(bool) v.Argument.Value;
+			}
+			var isKeyProp = attr.Properties.FirstOrDefault (p => p.Name == "IsKeyword");
+			var isKeyword = isKeyProp.Name != null && ((bool) isKeyProp.Argument.Value) == true;
+			var arrRankProp = attr.Properties.FirstOrDefault (p => p.Name == "ArrayRank");
+			if (arrRankProp.Name != null && arrRankProp.Argument.Value is int rank) {
+				r.Name = new string ('[', rank) + (isKeyword ? r.Name : "L" + r.Name + ";");
+			}
+		}
+		return r;
+	}
+
+	internal static RegisterAttribute? ToRegisterAttribute (CustomAttribute attr)
+	{
+		// attr.Resolve ();
+		RegisterAttribute? r = null;
+		if (attr.ConstructorArguments.Count == 1)
+			r = new RegisterAttribute ((string) attr.ConstructorArguments [0].Value, attr);
+		else if (attr.ConstructorArguments.Count == 3)
+			r = new RegisterAttribute (
+					(string) attr.ConstructorArguments [0].Value,
+					(string) attr.ConstructorArguments [1].Value,
+					(string) attr.ConstructorArguments [2].Value,
+					attr);
+		if (r != null) {
+			var v = attr.Properties.FirstOrDefault (p => p.Name == "DoNotGenerateAcw");
+			r.DoNotGenerateAcw = v.Name == null ? false : (bool) v.Argument.Value;
```

Comment:
Does something need to check if `v` is null? Maybe we should add `#nullable enable` to the top of this file.

Reply:
Good thought, I added `#nullable enable` locally and there are no warnings, so I will commit it with a future PR. 

Here, `var` is the `struct` `CustomAttributeNamedArgument` so the "Default" will be an empty `struct` rather than `null`.